### PR TITLE
Terms and Conditions Component

### PR DESCRIPTION
**Pull request by barshat7**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes and the overall coverage did not decrease.
- [ ] All unit tests pass on CircleCi.
- [ ] I ran all tests locally and they pass.

### CODE REVIEW

**Comment by rismehta** [comment]
There are test failures

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/TermsAndConditions.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+
+package com.adobe.cq.forms.core.components.models.form;
+
+import org.osgi.annotation.versioning.ConsumerType;
+
+@ConsumerType
+public interface TermsAndConditions extends Container, ContainerConstraint {
+
+    boolean isShowApprovalOption();
```
Comment (rismehta) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/TermsAndConditions.java:24]: All interfaces should have detailed documentation

Comment (rismehta) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/TermsAndConditions.java:24]: Please bump the minor version of package-info

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v1/form/TermsAndConditionsImpl.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    @JsonIgnore
+    @ValueMapValue(injectionStrategy = InjectionStrategy.OPTIONAL)
+    @Default(booleanValues = false)
+    private boolean showAsPopup;
+
+    @ValueMapValue(injectionStrategy = InjectionStrategy.OPTIONAL, name = "wrapData")
+    @Default(booleanValues = true)
+    @JsonIgnore
+    protected boolean wrapData;
```
Comment (rismehta) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v1/form/TermsAndConditionsImpl.java:70]: Why is this here, it should come from base class ?

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v1/form/TermsAndConditionsImpl.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    @ValueMapValue(injectionStrategy = InjectionStrategy.OPTIONAL)
+    protected boolean dorExclusion;
```
Comment (rismehta) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v1/form/TermsAndConditionsImpl.java:73]: Why is this specifically added here, it should come from base ?

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v1/form/TermsAndConditionsImpl.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        customDorProperties.put("dorExclusion", dorExclusion);
+        return customDorProperties;
+    }
+
+    @Override
+    protected List<Resource> getFilteredChildrenResources() {
+        List<Resource> childResources = getFilteredChildrenResources(resource);
+        // the tnc component will either have links or consent text based upon showLink value
+        if (showLink) {
+            childResources.removeIf(child -> FieldType.PLAIN_TEXT.getValue().equals(child.getValueMap().get("fieldType")));
```
Comment (rismehta) [inline bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/models/v1/form/TermsAndConditionsImpl.java:126]: please create a constant for fieldType and re-use the same.

File: bundles/af-core/src/test/resources/form/termsandconditions/test-content.json [hunk 5]
```diff
@@ -0,0 +1,87 @@
+{
+  "termsandconditions": {
+    "showApprovalOption": "true",
+    "jcr:title": "Terms And Conditions",
+    "fd:tnc": true,
+    "showAsPopup": "true",
```
Comment (rismehta) [inline bundles/af-core/src/test/resources/form/termsandconditions/test-content.json:6]: These should be boolean instead of string, similarly showApprovalOption should be boolean

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/termsandconditions/v1/termsandconditions/clientlibs/runtime/js/termsandconditionsview.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        }
+
+        #handleScroll() {
+            const intersection = this.element.querySelector('.cmp-adaptiveform-termsandcondition__text-intersect');
+            if (intersection) {
+                const self = this;
+                const onIntersection = ([{isIntersecting}]) => {
+                    if (isIntersecting) {
+                        const approvalCheckbox = self.getModel()._children.filter(child => child.fieldType === 'checkbox')[0];
+                        approvalCheckbox.dispatch(new Event('custom:scrollDone'));
```
Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/termsandconditions/v1/termsandconditions/clientlibs/runtime/js/termsandconditionsview.js:64]: Once intersection is found, you also need to unobserve on the target element, else there would be a memory leak

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/termsandconditions/v1/termsandconditions/_cq_template.xml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    jcr:primaryType="nt:unstructured"
+    jcr:title="I agree to the terms &amp; conditions"
+    sling:resourceType="core/fd/components/form/checkbox/v1/checkbox"
+    enabled="false"
+    checkedValue="true"
+    fieldType="checkbox">
+    <fd:events
+      jcr:primaryType="nt:unstructured"
+      custom_scrollDone="\{enabled: true()}"
+      custom_scrollInitialized="\{enabled: false()}"/>
```
Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/termsandconditions/v1/termsandconditions/_cq_template.xml:27]: This event is never thown, you use the initialize event to set the default state

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/termsandconditions/v1/termsandconditions/_cq_template.xml [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    name="approvalcheckbox"
+    jcr:primaryType="nt:unstructured"
+    jcr:title="I agree to the terms &amp; conditions"
+    sling:resourceType="core/fd/components/form/checkbox/v1/checkbox"
+    enabled="false"
+    checkedValue="true"
+    fieldType="checkbox">
+    <fd:events
+      jcr:primaryType="nt:unstructured"
+      custom_scrollDone="\{enabled: true()}"
```
Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/termsandconditions/v1/termsandconditions/_cq_template.xml:26]: Why is this exposed via rule editor, shouldn't we hardcode in the view when view itself is dispatching this event ? If someone forgets this while creating proxy, the whole TnC would be broken ? I would preview to do this via subscribe in the view layer itself

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/termsandconditions/v1/termsandconditions/clientlibs/editor/js/termsandconditions.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+/* global jQuery */
+(function($) {
+    "use strict";
+
+    var Utils = window.CQ.FormsCoreComponents.Utils.v1,
+        EDIT_DIALOG = ".cmp-adaptiveform-termsandconditions__editdialog";
+
+    function handleToggleOfApprovalOptions(dialog) {
+        var approvalOption = dialog.find('coral-checkbox[name="./showApprovalOption"]')[0];
```
Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/termsandconditions/v1/termsandconditions/clientlibs/editor/js/termsandconditions.js:24]: Please don't use coral elements, access these via granite class which are BEM based

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/termsandconditions/v1/termsandconditions/README.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+```
+
+
+
+## JavaScript Data Attribute Bindings
+
+The following attributes must be added for the initialization of the text-input component in the form view:
+1. `data-cmp-is="adaptiveFormTermsAndConditions"`
+2. `data-cmp-adaptiveformcontainer-path="${formstructparser.formContainerPath}"`
+3. `data-cmp-visible` having a boolean value to indicate whether the field is currently visible or not
```
Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/termsandconditions/v1/termsandconditions/README.md:78]: You need to also add data-cmp-required and data-cmp-readonly attributes

File: ui.tests/test-module/specs/termsandconditions/tnc.runtime.spec.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            model.visible = false;
+            return checkHTML(model.id, model.getState(), tabView, count);
+        }).then(() => {
+            model.enable = false;
+            return checkHTML(model.id, model.getState(), tabView, count);
+        });
+    });
+
+
+    it("OOTB behaviour -> should enable approval checkbox only if links visited", () => {
```
Comment (rismehta) [inline ui.tests/test-module/specs/termsandconditions/tnc.runtime.spec.js:72]: Please add runtime test cases for different states,
1. popup
2. WrapData
3. showLink

**Review by rismehta — CHANGES_REQUESTED** [review]
check comments

File: bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/TermsAndConditions.java (hunk 1 above)
Comment (rismehta) [reply bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/models/form/TermsAndConditions.java:24]: Please add documentation for JS in the view layer too

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/termsandconditions/v1/termsandconditions/clientlibs/runtime/js/termsandconditionsview.js (hunk 6 above)
Comment (barshat7) [reply ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/termsandconditions/v1/termsandconditions/clientlibs/runtime/js/termsandconditionsview.js:67]: Done

File: ui.tests/test-module/specs/termsandconditions/tnc.runtime.spec.js (hunk 11 above)
Comment (barshat7) [reply ui.tests/test-module/specs/termsandconditions/tnc.runtime.spec.js:72]: Done
